### PR TITLE
#753 Added application master password.

### PR DESCRIPTION
**Pull request by vldmrrr**
Implementation of feature requested in issue #753 - password protection.

### CODE REVIEW

File: k9mail/src/main/res/layout/master_password_prompt.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    <!-- Intro text -->
+    <TextView
+      android:id="@+id/master_password_prompt_intro"
+      android:textAppearance="?android:attr/textAppearanceMedium"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:layout_marginBottom="15dip"/>
+
+    <!-- Password prompt for the incoming server. Won't be shown for accounts without
```
Comment (philipwhiuk) [inline k9mail/src/main/res/layout/master_password_prompt.xml]: Inaccurate comment

File: k9mail/src/main/res/layout/master_password_prompt.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	      android:layout_height="wrap_content"/>
+	    <EditText
+	      android:id="@+id/master_password_input"
+	      android:inputType="textPassword"
+	      android:layout_height="wrap_content"
+	      android:layout_width="match_parent"
+	      android:layout_marginBottom="10dip"/>
+    </LinearLayout>
+
+    <!-- Password prompt for the outgoing server. Won't be shown for WebDAV accounts,
```
Comment (philipwhiuk) [inline k9mail/src/main/res/layout/master_password_prompt.xml]: Inaccurate comment

File: k9mail/src/main/java/com/fsck/k9/MasterPassword.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            final ScrollView scrollView = new ScrollView(activity);
+            final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
+            builder.setTitle(R.string.master_password_dialog_title);
+            builder.setView(scrollView);
+            builder.setPositiveButton(R.string.okay_action,
+                new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
+                        String password = mPasswordView.getText().toString();
+                        if (password == K9.getMasterPassword()) {
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/MasterPassword.java]: String comparison using ==

File: k9mail/src/main/java/com/fsck/k9/MasterPassword.java [hunk 4]
```diff
@@ -29,6 +33,25 @@ private void work(Activity activity) {
...
+                md.update(clearText.getBytes("UTF-8"));
+                byte[] bytes = md.digest();
+                Formatter f = new Formatter();
+                for (byte b : bytes)
+                    f.format("%02x", b);
+                String res = f.toString();
+                f.close();
+                return res;
+            } catch (NoSuchAlgorithmException e) {
+                e.printStackTrace();
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/MasterPassword.java]: Use logging, don't print to standard out.

File: k9mail/src/main/java/com/fsck/k9/MasterPassword.java [hunk 5]
```diff
@@ -29,6 +33,25 @@ private void work(Activity activity) {
...
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/MasterPassword.java]: Ditto

File: k9mail/src/main/java/com/fsck/k9/MasterPassword.java [hunk 6]
```diff
@@ -29,6 +33,25 @@ private void work(Activity activity) {
         (new PasswordDialog()).show(activity);
     }
 
+    public static String hash(String clearText) {
+            try {
+                MessageDigest md = MessageDigest.getInstance("MD5");
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/MasterPassword.java]: MD5 is broken, please don't use it - even if this is basically just for show.

**Comment by Valodim** [comment]
if we do a master password thing we should really encrypt the database with it, it's not really very useful to have otherwise.

**Comment by vldmrrr** [comment]
The goal of proposed patch is to provide control of access to user interface of the application, not to the application data. Encryption of database shall be addressed as a separate feature.

**Comment by Valodim** [comment]
Can we do this the other way around? I feel like this might give users a false sense of security :\

**Comment by cketti** [comment]
I'm not a big fan of application passwords. Instead my advice is to set up a device password and enable device encryption. Offering locking and encryption on the application-level only adds very little additional security while it significantly complicates the app architecture.

Even only locking the UI of the app is a rabbit hole. Because then you'll also want to protect notification actions, probably also the notification content. I'd rather not go there.

**Pull request closed without merging**